### ISSUE TITLE
Floating-point precision

### ISSUE THREAD

**Issue by pzp1997** [mdgriffith/elm-color#1]
I don't really like that the RGB components are stored as floats. It seems like we're introducing potential imprecision unnecessarily. Maybe this is less of an issue in practice than I'm making it out to be though.

**Comment by mdgriffith** [comment]
Good point.  Originally I had the rgb channels as `Int` between 0 and 255 as we're used to.  @kuon brought up the point that 0-255 was potentially unnecessarily constrained.

So, here's my breakdown of various tradeoffs after doing some color research:

It's only sorta tangentially related to the floats, I got excited researching color :)

**8bit color** (256 values per channel)
     - Nearly all monitors display 8 bit color

**10bit color** (1024 values per channel)
    - Some professional monitors can display 10bit color
    - Can be nice for processing/manipulating colors because of the extra information.

Also of note, apparently `rgb()` definitions in css can accept either an int in the range of 0-255 or a percentage as a float: https://developer.mozilla.org/en-US/docs/Web/CSS/color_value

I actually didn't know about the percentage part.

So, basically the trade off is either defining an absolute range like (0-255) or (0-1023), etc., or defining this as normalized floats between 0 and 1.0.

My thinking is that `elm-color` is meant to define a single type that packages can use to communicate a color, though other packages don't actually have to process things as floats.

So, you could have `superAccurateColorCalculator5000 : Config -> Color -> Color`, which converts every channel to `Int` at the color accuracy you want to process in (i.e. to 0-1023), do the processing, and then convert back to normalized 0-1.0 range.  I think you'd avoid meaningful floating point issues that way.

**Comment by kuon** [comment]
float has much more precision that constrained int in the 0-1 range.

If you do gpu programming, you will see that normalized floats are the way to go to store colors and do math with them, and in js floats are 64bit, so we are good precision wise. In fact, even int are floats in js iirc so we are not wasting memory.

I really think it is the sanest things to do, at least for storage. 

On March 30, 2018 4:43:30 PM GMT+02:00, Matthew Griffith <notifications@github.com> wrote:
>Good point.  Originally I had the rgb channels as `Int` between 0 and
>255 as we're used to.  @kuon brought up the point that 0-255 was
>potentially unnecessarily constrained.
>
>So, here's my breakdown of various tradeoffs after doing some color
>research:
>
>It's only sorta tangentially related to the floats, I got excited
>researching color :)
>
>**8bit color** (256 values per channel)
>     - Nearly all monitors display 8 bit color
>
>**10bit color** (1024 values per channel)
>    - Some professional monitors can display 10bit color
>- Can be nice for processing/manipulating colors because of the extra
>information.
>
>Also of note, apparently `rgb()` definitions in css can accept either
>an int in the range of 0-255 or a percentage as a float:
>https://developer.mozilla.org/en-US/docs/Web/CSS/color_value
>
>I actually didn't know about the percentage part.
>
>So, basically the trade off is either defining an absolute range like
>(0-255) or (0-1023), etc., or defining this as normalized floats
>between 0 and 1.0.
>
>My thinking is that `elm-color` is meant to define a single type that
>packages can use to communicate a color, though other packages don't
>actually have to process things as floats.
>
>So, you could have `superAccurateColorCalculator5000 : Config -> Color
>-> Color`, which converts every channel to `Int` at the color accuracy
>you want to process in (i.e. to 0-1023), do the processing, and then
>convert back to normalized 0-1.0 range.  I think you'd avoid meaningful
>floating point issues that way.
>
>
>
>
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/mdgriffith/elm-color/issues/1#issuecomment-377538909

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
